src/main/java/seedu/tasklist/model/ModelManager.java

```diff
@@ -73,20 +83,50 @@ public synchronized void deleteTask(ReadOnlyTask target) throws TaskNotFoundExce
         taskList.removeTask(target);
         updateFilteredListToShowIncomplete();
         indicateTaskListChanged();
+        addToUndoStack(UndoCommand.DEL_CMD_ID, (Task)target);
     }
 
     @Override
     public synchronized void addTask(Task task) throws UniqueTaskList.DuplicateTaskException {
         taskList.addTask(task);
         updateFilteredListToShowIncomplete();
         indicateTaskListChanged();
+        addToUndoStack(UndoCommand.ADD_CMD_ID, task);
+    }
+    
+    @Override
+    public synchronized void updateTask(Task taskToUpdate, TaskDetails taskDetails, StartTime startTime, EndTime endTime, Priority priority, UniqueTagList tags) throws UniqueTaskList.DuplicateTaskException {
+        taskList.updateTask(taskToUpdate, taskDetails, startTime, endTime, priority);
+        updateFilteredListToShowIncomplete();
+        indicateTaskListChanged();
+        Task originalTask = new Task(taskDetails, startTime, endTime, priority, tags);
+        addToUndoStack(UndoCommand.UPD_CMD_ID, taskToUpdate, originalTask);
     }
     
     @Override
     public synchronized void markTaskAsComplete(ReadOnlyTask task) throws TaskNotFoundException {
         taskList.markTaskAsComplete(task);
         updateFilteredListToShowIncomplete();
         indicateTaskListChanged();
+        addToUndoStack(UndoCommand.DONE_CMD_ID, (Task)task);
+    }
+    
+    @Override
+    public synchronized void markTaskAsIncomplete(ReadOnlyTask task) throws TaskNotFoundException {
+        taskList.markTaskAsIncomplete(task);
+        updateFilteredListToShowIncomplete();
+        indicateTaskListChanged();
+        addToUndoStack(UndoCommand.DONE_CMD_ID, (Task)task);
+    }
+
+    private void addToUndoStack(int undoID, Task... tasks) {
+        if (undoStack.size()==3){
+            undoStack.remove(undoStack.size()-1);
+        }
+        else{
```

Comment:
Should not be an else statement. If stack size is 3, we remove 1 elem, then add the new one in. With the else statement, if stack size is 3, we remove 1 elem and stop.